### PR TITLE
feat: Upload support customize previewFile

### LINKED ISSUE #15885: Uploading a big image file is freezing an app for few seconds
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/ywn9lorknx)

### Steps to reproduce
Try to upload image file with 20mb size

### What is expected?
The app should work as before

### What is actually happening?
The app is freezing for 4 seconds (between 10 and 14 seconds on the screenshot) on my mac (i5-7360U 2.3GHz)

| Environment | Info |
|---|---|
| antd | 3.16.1 |
| React | 16.8.4 |
| System | macOS Mojave 10.14.2 (18C54) |
| Browser | Chrome Version 72.0.3626.121 (Official Build) (64-bit) |

---
It happens because Upload component tries to make thumbnail from image file. It calls FileReader for it and if file is big it will freeze the app
I think we should add size check, for example don't make thumbnail preview if image size is bigger than 10mb.

If it ok, I can do PR

chrome performance screenshot

### PR DESCRIPTION
**Pull request by zombieJ**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### 👻 What's the background?

close #15885 
Laggy when upload large image file

### 💡 Solution

Provides `previewFile` to customize preview logic.

### 📝 Changelog

Upload support customize previewFile

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed
- [ ] Test ready

-----
[View rendered components/upload/demo/preview-file.md](https://github.com/ant-design/ant-design/blob/upload-preview-file/components/upload/demo/preview-file.md)
[View rendered components/upload/index.en-US.md](https://github.com/ant-design/ant-design/blob/upload-preview-file/components/upload/index.en-US.md)
[View rendered components/upload/index.zh-CN.md](https://github.com/ant-design/ant-design/blob/upload-preview-file/components/upload/index.zh-CN.md)

### CODE REVIEW

File: components/upload/Upload.tsx [hunk 1]
```diff
@@ -238,12 +238,13 @@ class Upload extends React.Component<UploadProps, UploadState> {
...
 
   renderUploadList = (locale: UploadLocale) => {
-    const { showUploadList, listType, onPreview } = this.props;
+    const { showUploadList, listType, onPreview, previewFile } = this.props;
     const { showRemoveIcon, showPreviewIcon } = showUploadList as any;
     return (
       <UploadList
         listType={listType}
         items={this.state.fileList}
+        previewFile={previewFile}
```
Comment (afc163) [inline components/upload/Upload.tsx]: previewImage

File: components/upload/UploadList.tsx [hunk 2]
```diff
@@ -64,6 +64,13 @@ export default class UploadList extends React.Component<UploadListProps, any> {
 
   // https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL
   previewFile = (file: File | Blob, callback: Function) => {
+    const { previewFile } = this.props;
+
+    if (previewFile) {
+      previewFile(file, callback);
+      return;
+    }
+
```
Comment (afc163) [inline components/upload/UploadList.tsx]: We should remove `readAsDataURL` from here, move it into `previewFile` demo.

File: components/upload/demo/preview-file.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+````jsx
+import { Upload, Button, Icon } from 'antd';
+
+const props = {
+  action: '//jsonplaceholder.typicode.com/posts/',
+  listType: 'picture',
+  previewFile(file, callback) {
+    // Your process logic. Here we just mock to the same file
+    callback('https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png');
```
Comment (afc163) [inline components/upload/demo/preview-file.md]: how about accepting a string or a promise?

File: components/upload/Upload.tsx (hunk 1 above)
Comment (zombieJ) [reply components/upload/Upload.tsx]: File can be video or other non-image type. `previewFile` may better than image.

File: components/upload/UploadList.tsx (hunk 2 above)
Comment (zombieJ) [reply components/upload/UploadList.tsx]: This will be break change. Move it as `previewFile` default prop may better.

Comment (afc163) [reply components/upload/UploadList.tsx]: Try this https://stackoverflow.com/a/25813769

File: components/upload/demo/preview-file.md [hunk 4]
```diff
@@ -19,9 +19,10 @@ import { Upload, Button, Icon } from 'antd';
 const props = {
   action: '//jsonplaceholder.typicode.com/posts/',
   listType: 'picture',
-  previewFile(file, callback) {
+  previewFile(file) {
+    console.log('Your upload file:', file);
     // Your process logic. Here we just mock to the same file
-    callback('https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png');
+    return Promise.resolve('https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png');
```
Comment (afc163) [inline components/upload/demo/preview-file.md]: 搞一个 fetch 做例子。

Comment (afc163) [reply components/upload/demo/preview-file.md]: https://next.json-generator.com

**Comment by zombieJ** [comment]
抽了 `spyElementPrototype` / `spyElementPrototypes` 用于支持原生 HTMLElement 的 prototype spy:
```js
// property type
const imageSpy = spyElementPrototype(Image, 'src', {
  set() {},
});

// function type
const canvasSpy = spyElementPrototype(HTMLCanvasElement,  'toDataURL', () => null);

// Restore
imageSpy.mockRestore();
canvasSpy.mockRestore();
```

合完后，把其他手工的也换成封装的。

**Comment by zombieJ** [comment]
Done. Please check.

**Review by afc163 — APPROVED** [review]

**Comment by wuxingjiang007** [comment]
用不了啊，也没报错, tslint提示：This may be converted to an async function.

**Pull request merged**